core/solver/ir.cpp

```diff
@@ -0,0 +1,121 @@
+
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <ginkgo/core/solver/ir.hpp>
+
+
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+#include "core/solver/ir_kernels.hpp"
+
+
+namespace gko {
+namespace solver {
+namespace ir {
+
+
+GKO_REGISTER_OPERATION(initialize, ir::initialize);
+
+
+}  // namespace ir
+
+
+template <typename ValueType>
+void Ir<ValueType>::apply_impl(const LinOp *b, LinOp *x) const
+{
+    using Vector = matrix::Dense<ValueType>;
+    constexpr uint8 relative_stopping_id{1};
+
+    auto exec = this->get_executor();
+    auto one_op = initialize<Vector>({one<ValueType>()}, exec);
+    auto neg_one_op = initialize<Vector>({-one<ValueType>()}, exec);
+
+    auto dense_b = as<const Vector>(b);
+    auto dense_x = as<Vector>(x);
+    auto residual = Vector::create_with_config_of(dense_b);
+
+    bool one_changed{};
+    Array<stopping_status> stop_status(exec, dense_b->get_size()[1]);
+    exec->run(ir::make_initialize(&stop_status));
+
+    residual->copy_from(dense_b);
+    system_matrix_->apply(lend(neg_one_op), dense_x, lend(one_op),
+                          lend(residual));
+
+    auto stop_criterion = stop_criterion_factory_->generate(
+        system_matrix_, std::shared_ptr<const LinOp>(b, [](const LinOp *) {}),
+        x, lend(residual));
+
+    int iter = -1;
+    while (true) {
+        ++iter;
+        this->template log<log::Logger::iteration_complete>(
+            this, iter, lend(residual), dense_x);
+
+        if (stop_criterion->update()
+                .num_iterations(iter)
+                .residual(lend(residual))
+                .solution(dense_x)
+                .check(relative_stopping_id, true, &stop_status,
+                       &one_changed)) {
+            break;
+        }
+
+        solver_->apply(lend(one_op), lend(residual), lend(one_op), dense_x);
+        residual->copy_from(dense_b);
+        system_matrix_->apply(lend(neg_one_op), dense_x, lend(one_op),
```

Comment:
It's quite simple, but nonetheless I believe we in general put as comments the general algorithm of the solver, maybe you could also do that here? (or was that only for the step kernels?)

Reply:
it was only for the step kernels